modules/ml/src/svmsgd.cpp

```diff
@@ -0,0 +1,543 @@
+// or tort (including negligence or otherwise) arising in any way out of
+// the use of this software, even if advised of the possibility of such damage.
+//
+//M*/
+
+#include "precomp.hpp"
+#include "limits"
+
+#include <iostream>
+
+using std::cout;
+using std::endl;
+
+/****************************************************************************************\
+*                        Stochastic Gradient Descent SVM Classifier                      *
+\****************************************************************************************/
+
+namespace cv
+{
+namespace ml
+{
+
+class SVMSGDImpl : public SVMSGD
+{
+
+public:
+    SVMSGDImpl();
+
+    virtual ~SVMSGDImpl() {}
+
+    virtual bool train(const Ptr<TrainData>& data, int);
+
+    virtual float predict( InputArray samples, OutputArray results=noArray(), int flags = 0 ) const;
+
+    virtual bool isClassifier() const;
+
+    virtual bool isTrained() const;
+
+    virtual void clear();
+
+    virtual void write(FileStorage &fs) const;
+
+    virtual void read(const FileNode &fn);
+
+    virtual Mat getWeights(){ return weights_; }
+
+    virtual float getShift(){ return shift_; }
+
+    virtual int getVarCount() const { return weights_.cols; }
+
+    virtual String getDefaultName() const {return "opencv_ml_svmsgd";}
+
+    virtual void setOptimalParameters(int svmsgdType = ASGD, int marginType = SOFT_MARGIN);
+
+    CV_IMPL_PROPERTY(int, SvmsgdType, params.svmsgdType)
+    CV_IMPL_PROPERTY(int, MarginType, params.marginType)
+    CV_IMPL_PROPERTY(float, MarginRegularization, params.marginRegularization)
+    CV_IMPL_PROPERTY(float, InitialStepSize, params.initialStepSize)
+    CV_IMPL_PROPERTY(float, StepDecreasingPower, params.stepDecreasingPower)
+    CV_IMPL_PROPERTY_S(cv::TermCriteria, TermCriteria, params.termCrit)
+
+private:
+    void updateWeights(InputArray sample, bool isPositive, float stepSize, Mat &weights);
+
+    std::pair<bool,bool> areClassesEmpty(Mat responses);
+
+    void writeParams( FileStorage &fs ) const;
+
+    void readParams( const FileNode &fn );
+
+    static inline bool isPositive(float val) { return val > 0; }
+
+    static void normalizeSamples(Mat &matrix, Mat &average, float &multiplier);
+
+    float calcShift(InputArray _samples, InputArray _responses) const;
+
+    static void makeExtendedTrainSamples(const Mat &trainSamples, Mat &extendedTrainSamples, Mat &average, float &multiplier);
+
+
+
+    // Vector with SVM weights
+    Mat weights_;
+    float shift_;
+
+    // Parameters for learning
+    struct SVMSGDParams
+    {
+        float marginRegularization;
+        float initialStepSize;
+        float stepDecreasingPower;
+        TermCriteria termCrit;
+        int svmsgdType;
+        int marginType;
+    };
+
+    SVMSGDParams params;
+};
+
+Ptr<SVMSGD> SVMSGD::create()
+{
+    return makePtr<SVMSGDImpl>();
+}
+
+std::pair<bool,bool> SVMSGDImpl::areClassesEmpty(Mat responses)
```

Comment:
I wonder, if something like `any(responses > 0) && any(responses < 0)` would do the thing